presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java

```diff
@@ -0,0 +1,210 @@
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.mergeJoin;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static io.airlift.tpch.TpchTable.CUSTOMER;
+import static io.airlift.tpch.TpchTable.LINE_ITEM;
+import static io.airlift.tpch.TpchTable.NATION;
+import static io.airlift.tpch.TpchTable.ORDERS;
+
+public class TestMergeJoinPlan
+        extends AbstractTestQueryFramework
+{
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        return HiveQueryRunner.createQueryRunner(
+                ImmutableList.of(ORDERS, LINE_ITEM, CUSTOMER, NATION),
+                ImmutableMap.of(),
+                Optional.empty());
+    }
+
+    @Test
+    public void testSessionProperty()
+    {
+        QueryRunner queryRunner = getQueryRunner();
+
+        try {
+            queryRunner.execute("CREATE TABLE test_join_customer WITH ( \n" +
+                    "  bucket_count = 4, bucketed_by = ARRAY['custkey'], \n" +
+                    "  sorted_by = ARRAY['custkey'], partitioned_by=array['ds']) AS \n" +
+                    "SELECT *, '2021-07-11' as ds FROM tpch.sf1.customer LIMIT 1000");
+
+            queryRunner.execute("CREATE TABLE test_join_order WITH ( \n" +
+                    "  bucket_count = 4, bucketed_by = ARRAY['custkey'], \n" +
+                    "  sorted_by = ARRAY['custkey'], partitioned_by=array['ds']) AS \n" +
+                    "SELECT *, '2021-07-11' as ds FROM tpch.sf1.\"orders\" LIMIT 1000");
+
+            // By default, we don't enable merge join
+            assertPlan(
+                    "select * from test_join_customer join test_join_order on test_join_customer.custkey = test_join_order.custkey",
+                    joinPlan("test_join_customer", "test_join_order", ImmutableList.of("custkey"), ImmutableList.of("custkey"), false));
+
+            // When merge join is preferred and data properties requirements for merge join are met
+            assertPlan(
+                    mergeJoinEnabled(),
+                    "select * from test_join_customer join test_join_order on test_join_customer.custkey = test_join_order.custkey",
+                    joinPlan("test_join_customer", "test_join_order", ImmutableList.of("custkey"), ImmutableList.of("custkey"), true));
+        }
+        finally {
+            queryRunner.execute("DROP TABLE IF EXISTS test_join_customer");
+            queryRunner.execute("DROP TABLE IF EXISTS test_join_order");
+        }
+    }
+
+    @Test
+    public void testDifferentBucketedByKey()
+    {
+        QueryRunner queryRunner = getQueryRunner();
+
+        try {
+            queryRunner.execute("CREATE TABLE test_join_customer2 WITH ( \n" +
+                    "  bucket_count = 4, bucketed_by = ARRAY['name'], \n" +
```

Comment:
both
1. because it's not bucketed on the same key as the other table, two scans can't be in the same stage, there would be Exchange
2. If they're not bucketed on join key, two scans can't be in the same stage either